src/details/ArborX_DetailsBruteForceImpl.hpp

```diff
@@ -53,26 +53,94 @@ struct BruteForceImpl
 
     int const n_primitives = AccessPrimitives::size(primitives);
     int const n_predicates = AccessPredicates::size(predicates);
+    int max_scratch_size = TeamPolicy::scratch_size_max(0) / 2;
+    int const predicates_per_team = max_scratch_size / sizeof(PredicateType);
+    int const primitives_per_team = max_scratch_size / sizeof(PrimitiveType);
+
+    int const n_primitive_tiles =
+        ceil((float)n_primitives / primitives_per_team);
+    int const n_predicate_tiles =
+        ceil((float)n_predicates / predicates_per_team);
+    int const n_teams = n_primitive_tiles * n_predicate_tiles;
+
+    using ScratchPredicateType =
+        Kokkos::View<PredicateType *,
+                     typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ScratchPrimitiveType =
+        Kokkos::View<PrimitiveType *,
+                     typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    int scratch_size = ScratchPredicateType::shmem_size(predicates_per_team) +
+                       ScratchPrimitiveType::shmem_size(primitives_per_team);
 
     Kokkos::parallel_for(
         "ArborX::BruteForce::query::spatial::"
         "check_all_predicates_against_all_primitives",
-        Kokkos::MDRangePolicy<ExecutionSpace, Kokkos::Rank<2>>(
-            space, {0, 0}, {n_primitives, n_predicates}),
-        KOKKOS_LAMBDA(int primitive_index, int predicate_index) {
-          auto const &predicate =
-              AccessPredicates::get(predicates, predicate_index);
-          auto const &primitive =
-              AccessPrimitives::get(primitives, primitive_index);
-          if (predicate(primitive))
+        TeamPolicy((long)n_teams, Kokkos::AUTO)
+            .set_scratch_size(0, Kokkos::PerTeam(scratch_size)),
+        KOKKOS_LAMBDA(const typename TeamPolicy::member_type &teamMember) {
+          // select the tiles of predicates/primitives checked by each team
+          int predicate_start = predicates_per_team *
+                                (teamMember.league_rank() / n_primitive_tiles);
+          int primitive_start = primitives_per_team *
+                                (teamMember.league_rank() % n_primitive_tiles);
+
+          int predicates_in_this_team = KokkosExt::min(
+              predicates_per_team, n_predicates - predicate_start);
+          int primitives_in_this_team = KokkosExt::min(
+              primitives_per_team, n_primitives - primitive_start);
+
+          ScratchPredicateType scratch_predicates(teamMember.team_scratch(0),
+                                                  predicates_per_team);
+          ScratchPrimitiveType scratch_primitives(teamMember.team_scratch(0),
+                                                  primitives_per_team);
+          // rank 0 in each team fills the scratch space with the
+          // predicates / primitives in the tile
```

Comment:
Why not just `TeamVectorRange` (or two `ThreadVectorRange` with different teams as suggested below)?

Reply:
Also, note that `vector_size == 1` by default. So all the `ThreadVectorRamge` loops are executed sequentially (within each team and could be replaced by a plain `for` loop).
It would be interesting to see how larger vector sizes would affect performance.

Reply:
I don't know what's the impact on performance of any of your suggestions. Is it worth looking at it?

Reply:
Probably so